// gavel/lib/validators/text-diff.js
// Generated by CoffeeScript 1.9.2
var DiffMatchPatch, TextDiff, ValidationErrors, errors, type;

ValidationErrors = require('./validation-errors').ValidationErrors;

errors = require('../errors');

DiffMatchPatch = require('googlediff');

type = require('is-type');

TextDiff = (function() {
  function TextDiff(real, expected) {
    var outError;
    this.real = real;
    this.expected = expected;
    if (!type.string(this.real)) {
      outError = new errors.DataNotStringError('String validator real: input data is not string');
      outError['data'] = this.real;
      throw outError;
    }
    if (!type.string(this.expected)) {
      outError = new errors.DataNotStringError('String validator expected: input data is not string');
      outError['data'] = this.expected;
      throw outError;
    }
  }

  TextDiff.prototype.validate = function() {
    var dmp, e, patch, sanitizeSurrogatePairs;
    sanitizeSurrogatePairs = function(data) {
      return data.replace(/[\uD800-\uDBFF]/g, '').replace(/[\uDC00-\uDFFF]/g, '');
    };
    this.output = null;
    dmp = new DiffMatchPatch;
    try {
      patch = dmp.patch_make(this.real, this.expected);
      this.output = dmp.patch_toText(patch);
      return this.output;
    } catch (_error) {
      e = _error;
      if (e instanceof URIError) {
        patch = dmp.patch_make(sanitizeSurrogatePairs(this.real), sanitizeSurrogatePairs(this.expected));
        this.output = dmp.patch_toText(patch);
        return this.output;
      } else {
        throw e;
      }
    }
  };

  TextDiff.prototype.evaluateOutputToResults = function(data) {
    if (!data) {
      data = this.output;
    }
    if (!data) {
      return [];
    }
    return [
      {
        severity: 'error',
        message: 'Real and expected data does not match.'
      }
    ];
  };

  return TextDiff;

})();

module.exports = {
  TextDiff: TextDiff
};
