// dredd/lib/dredd-command.js
// Generated by CoffeeScript 1.10.0
var Dredd, DreddCommand, TERM_RETRY, TERM_TIMEOUT, configUtils, console, exports, fs, interactiveConfig, logger, optimist, parsePackageJson, path, spawn, spawnArgs, version;

path = require('path');

optimist = require('optimist');

console = require('console');

fs = require('fs');

spawnArgs = require('spawn-args');

spawn = require('child_process').spawn;

parsePackageJson = require('./parse-package-json');

Dredd = require('./dredd');

interactiveConfig = require('./interactive-config');

configUtils = require('./config-utils');

logger = require('./logger');

version = parsePackageJson(path.join(__dirname, '../package.json'));

TERM_TIMEOUT = 1000;

TERM_RETRY = 500;

DreddCommand = (function() {
  function DreddCommand(options, cb) {
    if (options == null) {
      options = {};
    }
    this.cb = cb;
    this.finished = false;
    this.exit = options.exit, this.custom = options.custom;
    this.serverProcessEnded = false;
    this.setExitOrCallback();
    if (this.custom == null) {
      this.custom = {};
    }
    if (!this.custom.cwd || typeof this.custom.cwd !== 'string') {
      this.custom.cwd = process.cwd();
    }
    if (!this.custom.argv || !Array.isArray(this.custom.argv)) {
      this.custom.argv = [];
    }
  }

  DreddCommand.prototype.setOptimistArgv = function() {
    this.optimist = optimist(this.custom['argv'], this.custom['cwd']);
    this.cliArgv = this.optimist.argv;
    this.optimist.usage("Usage:\n  $ dredd init\n\nOr:\n  $ dredd <path or URL to blueprint> <api_endpoint> [OPTIONS]\n\nExample:\n  $ dredd ./apiary.md http://localhost:3000 --dry-run").options(Dredd.options).wrap(80);
    return this.argv = this.optimist.argv;
  };

  DreddCommand.prototype.stopServer = function(callback) {
    var kill, start, term, timeout, waitForServerTermOrKill;
    if (this.serverProcess == null) {
      return callback();
    }
    term = (function(_this) {
      return function() {
        logger.info('Sending SIGTERM to the backend server');
        return _this.serverProcess.kill('SIGTERM');
      };
    })(this);
    kill = (function(_this) {
      return function() {
        logger.info('Killing backend server');
        return _this.serverProcess.kill('SIGKILL');
      };
    })(this);
    start = Date.now();
    term();
    waitForServerTermOrKill = (function(_this) {
      return function() {
        var timeout;
        if (_this.serverProcessEnded === true) {
          clearTimeout(timeout);
          return callback();
        } else {
          if ((Date.now() - start) < TERM_TIMEOUT) {
            term();
            return timeout = setTimeout(waitForServerTermOrKill, TERM_RETRY);
          } else {
            kill();
            clearTimeout(timeout);
            return callback();
          }
        }
      };
    })(this);
    return timeout = setTimeout(waitForServerTermOrKill, TERM_RETRY);
  };

  DreddCommand.prototype.setExitOrCallback = function() {
    if (!this.cb) {
      if (this.exit && (this.exit === process.exit)) {
        this.sigIntEventAdd = true;
      }
      if (this.exit) {
        return this._processExit = (function(_this) {
          return function(exitStatus) {
            _this.finished = true;
            return _this.stopServer(function() {
              return _this.exit(exitStatus);
            });
          };
        })(this);
      } else {
        return this._processExit = (function(_this) {
          return function(exitStatus) {
            return _this.stopServer(function() {
              return process.exit(exitStatus);
            });
          };
        })(this);
      }
    } else {
      return this._processExit = (function(_this) {
        return function(exitStatus) {
          _this.finished = true;
          if (_this.sigIntEventAdded) {
            if (_this.serverProcess != null) {
              _this.serverProcess.kill('SIGKILL');
            }
            process.removeEventListener('SIGINT', _this.commandSigInt);
          }
          _this.cb(exitStatus);
          return _this;
        };
      })(this);
    }
  };

  DreddCommand.prototype.moveBlueprintArgToPath = function() {
    if (!Array.isArray(this.argv['path'])) {
      return this.argv['path'] = this.argv['p'] = [this.argv['path']];
    }
  };

  DreddCommand.prototype.checkRequiredArgs = function() {
    var argError;
    argError = false;
    if (this.argv._[0] == null) {
      console.error("\nError: Must specify path to blueprint file.");
      argError = true;
    }
    if (this.argv._[1] == null) {
      console.error("\nError: Must specify api endpoint.");
      argError = true;
    }
    if (argError) {
      console.error("\n");
      this.optimist.showHelp(console.error);
      return this._processExit(1);
    }
  };

  DreddCommand.prototype.runExitingActions = function() {
    if (this.argv["_"][0] === "init" || this.argv.init === true) {
      this.finished = true;
      return interactiveConfig.run(this.argv, (function(_this) {
        return function(config) {
          configUtils.save(config);
          console.log("");
          console.log("Configuration saved to dredd.yml");
          console.log("");
          if (config['language'] === "nodejs") {
            console.log("Run test now, with:");
          } else {
            console.log("Install hooks handler and run Dredd test with:");
          }
          console.log("");
          if (config['language'] === 'ruby') {
            console.log("  $ gem install dredd_hooks");
          } else if (config['language'] === 'python') {
            console.log("  $ pip install dredd_hooks");
          } else if (config['language'] === 'php') {
            console.log("  $ composer require ddelnano/dredd-hooks-php --dev");
          } else if (config['language'] === 'perl') {
            console.log("  $ cpanm Dredd::Hooks");
          } else if (config['language'] === 'go') {
            console.log("  $ go get github.com/snikch/goodman");
          }
          console.log("  $ dredd");
          console.log("");
          return _this._processExit(0);
        };
      })(this));
    } else if (this.argv.help === true) {
      this.optimist.showHelp(console.error);
      return this._processExit(0);
    } else if (this.argv.version === true) {
      console.log(version);
      return this._processExit(0);
    }
  };

  DreddCommand.prototype.loadDreddFile = function() {
    var key, ref, results, value;
    if (fs.existsSync('./dredd.yml')) {
      logger.info('Configuration dredd.yml found, ignoring other arguments.');
      this.argv = configUtils.load();
    }
    ref = this.cliArgv;
    results = [];
    for (key in ref) {
      value = ref[key];
      if (key !== "_" && key !== "$0") {
        results.push(this.argv[key] = value);
      } else {
        results.push(void 0);
      }
    }
    return results;
  };

  DreddCommand.prototype.parseCustomConfig = function() {
    return this.argv.custom = configUtils.parseCustom(this.argv.custom);
  };

  DreddCommand.prototype.runServerAndThenDredd = function(callback) {
    var command, parsedArgs, waitMilis, waitSecs;
    if (this.argv['server'] == null) {
      return this.runDredd(this.dreddInstance);
    } else {
      parsedArgs = spawnArgs(this.argv['server']);
      command = parsedArgs.shift();
      this.serverProcess = spawn(command, parsedArgs);
      logger.info("Starting server with command: " + this.argv['server']);
      this.serverProcess.stdout.setEncoding('utf8');
      this.serverProcess.stdout.on('data', function(data) {
        return process.stdout.write(data.toString());
      });
      this.serverProcess.stderr.setEncoding('utf8');
      this.serverProcess.stderr.on('data', function(data) {
        return process.stdout.write(data.toString());
      });
      this.serverProcess.on('close', (function(_this) {
        return function(status) {
          _this.serverProcessEnded = true;
          if (status != null) {
            return logger.info('Backend server exited');
          } else {
            return logger.info('Backend server was killed');
          }
        };
      })(this));
      this.serverProcess.on('error', (function(_this) {
        return function(error) {
          logger.error("Server command failed, exiting Dredd...");
          return _this._processExit(2);
        };
      })(this));
      process.on('beforeExit', (function(_this) {
        return function() {
          if (_this.serverProcess != null) {
            return _this.serverProcess.kill('SIGKILL');
          }
        };
      })(this));
      process.on('exit', (function(_this) {
        return function() {
          if (_this.serverProcess != null) {
            return _this.serverProcess.kill('SIGKILL');
          }
        };
      })(this));
      waitSecs = parseInt(this.argv['server-wait'], 10);
      waitMilis = waitSecs * 1000;
      logger.info("Waiting " + waitSecs + " seconds for server command to start...");
      return this.wait = setTimeout((function(_this) {
        return function() {
          return _this.runDredd(_this.dreddInstance);
        };
      })(this), waitMilis);
    }
  };

  DreddCommand.prototype.run = function() {
    var configurationForDredd, e, error1, i, len, ref, task;
    ref = [this.setOptimistArgv, this.parseCustomConfig, this.runExitingActions, this.loadDreddFile, this.checkRequiredArgs, this.moveBlueprintArgToPath];
    for (i = 0, len = ref.length; i < len; i++) {
      task = ref[i];
      task.call(this);
      if (this.finished) {
        return;
      }
    }
    configurationForDredd = this.initConfig();
    this.dreddInstance = this.initDredd(configurationForDredd);
    try {
      this.runServerAndThenDredd();
    } catch (error1) {
      e = error1;
      logger.error(e.message);
      logger.error(e.stack);
      this.stopServer((function(_this) {
        return function() {
          return _this._processExit(2);
        };
      })(this));
    }
  };

  DreddCommand.prototype.lastArgvIsApiEndpoint = function() {
    this.server = this.argv._[this.argv._.length - 1];
    this.argv._.splice(this.argv._.length - 1, 1);
    return this;
  };

  DreddCommand.prototype.takeRestOfParamsAsPath = function() {
    this.argv['p'] = this.argv['path'] = this.argv['path'].concat(this.argv._);
    return this;
  };

  DreddCommand.prototype.initConfig = function() {
    var base, configuration;
    this.lastArgvIsApiEndpoint().takeRestOfParamsAsPath();
    configuration = {
      'server': this.server,
      'options': this.argv
    };
    if ((base = configuration.options).path == null) {
      base.path = [];
    }
    configuration.options.path.push(this.argv._[0]);
    configuration.custom = this.custom;
    return configuration;
  };

  DreddCommand.prototype.initDredd = function(configuration) {
    var dredd;
    dredd = new Dredd(configuration);
    return dredd;
  };

  DreddCommand.prototype.commandSigInt = function() {
    logger.error("\nShutting down from SIGINT (Ctrl-C)");
    return this.dreddInstance.transactionsComplete((function(_this) {
      return function() {
        return _this._processExit(0);
      };
    })(this));
  };

  DreddCommand.prototype.runDredd = function(dreddInstance) {
    if (this.sigIntEventAdd) {
      this.sigIntEventAdded = !(this.sigIntEventAdd = false);
      process.on('SIGINT', this.commandSigInt);
    }
    dreddInstance.run((function(_this) {
      return function(error, stats) {
        return _this.exitWithStatus(error, stats);
      };
    })(this));
    return this;
  };

  DreddCommand.prototype.exitWithStatus = function(error, stats) {
    if (error) {
      if (error.message) {
        logger.error(error.message);
      }
      return this._processExit(1);
    }
    if ((stats.failures + stats.errors) > 0) {
      this._processExit(1);
    } else {
      this._processExit(0);
    }
  };

  return DreddCommand;

})();

exports = module.exports = DreddCommand;
