// gavel/lib/errors.js
// Generated by CoffeeScript 1.9.2
var DataNotJsonParsableError, DataNotStringError, JsonSchemaNotValid, MalformedDataError, MalformedSchemaError, NotEnoughDataError, NotValidatableError, SchemaNotJsonParsableError, UnknownValidatorError,
  extend = function(child, parent) { for (var key in parent) { if (hasProp.call(parent, key)) child[key] = parent[key]; } function ctor() { this.constructor = child; } ctor.prototype = parent.prototype; child.prototype = new ctor(); child.__super__ = parent.prototype; return child; },
  hasProp = {}.hasOwnProperty;

MalformedDataError = (function(superClass) {
  extend(MalformedDataError, superClass);

  function MalformedDataError() {
    return MalformedDataError.__super__.constructor.apply(this, arguments);
  }

  return MalformedDataError;

})(Error);

DataNotJsonParsableError = (function(superClass) {
  extend(DataNotJsonParsableError, superClass);

  function DataNotJsonParsableError() {
    return DataNotJsonParsableError.__super__.constructor.apply(this, arguments);
  }

  return DataNotJsonParsableError;

})(MalformedDataError);

DataNotStringError = (function(superClass) {
  extend(DataNotStringError, superClass);

  function DataNotStringError() {
    return DataNotStringError.__super__.constructor.apply(this, arguments);
  }

  return DataNotStringError;

})(MalformedDataError);

MalformedSchemaError = (function(superClass) {
  extend(MalformedSchemaError, superClass);

  function MalformedSchemaError() {
    return MalformedSchemaError.__super__.constructor.apply(this, arguments);
  }

  return MalformedSchemaError;

})(Error);

SchemaNotJsonParsableError = (function(superClass) {
  extend(SchemaNotJsonParsableError, superClass);

  function SchemaNotJsonParsableError() {
    return SchemaNotJsonParsableError.__super__.constructor.apply(this, arguments);
  }

  return SchemaNotJsonParsableError;

})(MalformedSchemaError);

UnknownValidatorError = (function(superClass) {
  extend(UnknownValidatorError, superClass);

  function UnknownValidatorError() {
    return UnknownValidatorError.__super__.constructor.apply(this, arguments);
  }

  return UnknownValidatorError;

})(Error);

NotValidatableError = (function(superClass) {
  extend(NotValidatableError, superClass);

  function NotValidatableError() {
    return NotValidatableError.__super__.constructor.apply(this, arguments);
  }

  return NotValidatableError;

})(Error);

NotEnoughDataError = (function(superClass) {
  extend(NotEnoughDataError, superClass);

  function NotEnoughDataError() {
    return NotEnoughDataError.__super__.constructor.apply(this, arguments);
  }

  return NotEnoughDataError;

})(Error);

JsonSchemaNotValid = (function(superClass) {
  extend(JsonSchemaNotValid, superClass);

  function JsonSchemaNotValid() {
    return JsonSchemaNotValid.__super__.constructor.apply(this, arguments);
  }

  return JsonSchemaNotValid;

})(Error);

module.exports = {
  DataNotJsonParsableError: DataNotJsonParsableError,
  SchemaNotJsonParsableError: SchemaNotJsonParsableError,
  MalformedSchemaError: MalformedSchemaError,
  MalformedDataError: MalformedDataError,
  UnknownValidatorError: UnknownValidatorError,
  DataNotStringError: DataNotStringError,
  NotValidatableError: NotValidatableError,
  NotEnoughDataError: NotEnoughDataError,
  JsonSchemaNotValid: JsonSchemaNotValid
};
